### ISSUE TITLE
DAU/MAU is not counted correctly for OAuth relying parties

### ISSUE THREAD

**Issue by shane-tomlinson** [mozilla/fxa-auth-server#2957]
We calculate DAU/MAU based on the number of [certificate signs that occur](https://github.com/mozilla/fxa-auth-server/blob/7fe15bfc5ef4d89d7a4976a74f33340d1cc9aa05/lib/routes/sign.js#L1). 

Users of OAuth relying parties that make use of [long lived access tokens](https://bugzilla.mozilla.org/show_bug.cgi?id=1532438) or refresh tokens will only be counted towards MAU on the day the access was first granted, when the certificate was signed. Ongoing use of OAuth tokens will not count towards DAU/MAU.

DAU/MAU counting using certificate signs will be further reduced once [/authorization can be called with a sessionToken](https://github.com/mozilla/fxa-auth-server/pull/2932) and not an assertion, in this case no cert will be needed to get OAuth credentials. 

Possible solution:

* [Reduce the lifetime of an access token to 24 hours](https://bugzilla.mozilla.org/show_bug.cgi?id=1532438). 
* Log an event that counts towards DAU/MAU whenever [a refresh token is used to get an access token](https://github.com/mozilla/fxa-auth-server/blob/7fe15bfc5ef4d89d7a4976a74f33340d1cc9aa05/fxa-oauth-server/lib/routes/token.js#L90) in the /token endpoint of the oauth server.
* Log an event that counts towards DAU/MAU on calls to [POST /verify on the oauth server](https://github.com/mozilla/fxa-auth-server/blob/7fe15bfc5ef4d89d7a4976a74f33340d1cc9aa05/fxa-oauth-server/lib/routes/verify.js#L7) to handle relying parties that occasionally check the validity of a refresh or access token but do not request a new one.
* Add a `service` field whenever [emitting the `account.signed` event](https://github.com/mozilla/fxa-auth-server/blob/7fe15bfc5ef4d89d7a4976a74f33340d1cc9aa05/lib/routes/sign.js#L179) and only count towards DAU/MAU if `service=sync`.

I'm not sure how we go about associating a call to /token or /verify with the metrics context bundle that is associated with the login, we don't want to place a burden on reliers to propagate that info, @philbooth could probably chime in here. Maybe when an OAuth code is generated, the content server propagates the metrics context to the auth server and the metrics context is stashed in redis and associated with the corresponding OAuth code or access token (in the case of an implicit grant). When the code is traded for either an access token only, or an access token + refresh token, associate the metrics context with both token types.

cc @davismtl, @irrationalagent, @philbooth, @rfk, @clouserw

**Comment by philbooth** [comment]
> we don't want to place a burden on reliers to propagate that info

It's cool, we can do all that propagation work on our end.

**Comment by irrationalagent** [comment]
I don't have much advice to give on this right now (sorry). All I will say is that we should absolutely do this because we are leaving a bunch of MAU on the table with respect to non-sync reliers. And this unusable pile of funny money will only continue to accrue as more non-sync reliers are added. 

Also, if for whatever reason 24 hrs is too short, since its MAU that is currently more important than DAU (even though I think we as a team still care about it) the time-frame could be lengthened a bit without too much consequence to the metrics (it would still be much better than what we have now). But 24 would be ideal.

**Comment by rfk** [comment]
> Add a service field whenever emitting the account.signed event and only count
> towards DAU/MAU if service=sync.

It's not obvious to me why this one is necessary, since DAU/MAU are aggregated by unique uid anyway I don't think it matters if we count *both* an `account.signed` and some oauth-token-related event for the same activity.

Another thing to watch out for would be OAuth RPs that are using the tokens for their own purpoes (such as checking if their cache of the user's profile data is up-to-date) that does not actually correspond to activity from the user. Not entirely sure what to do about that though, apart from telling them not to do that :-)

(In an ideal world, each RP would be able to log RP-specific events into the shared ecosystem telemetry pipeline and we'd count DAU/MAU from there, but that seems like a longer-term thing)

**Comment by rfk** [comment]
> When the code is traded for either an access token only, or an access token + refresh token,
> associate the metrics context with both token types.

Do we need to associate metrics context with these tokens on an ongoing basis, or is it a once-of thing for the purposes of tracking flow completion?

**Comment by davismtl** [comment]
> Another thing to watch out for would be OAuth RPs that are using the tokens for their own purpoes (such as checking if their cache of the user's profile data is up-to-date) that does not actually correspond to activity from the user. Not entirely sure what to do about that though, apart from telling them not to do that :-)

Is it possible to have short-lived and long-lived tokens to split up responsibilities? For example, authentication could be "short"-er at like X days. This would force them to check with us regularly. Whereas, getting profile information could be a long-lived token at ~90 days, where if the user hasn't used the OAuth RP for that amount of time, they can no longer refresh the user data. 🤷🏼‍♂️

I'm just thinking out loud here. I'm pushing the limits of my OAuth knowledge.

**Comment by shane-tomlinson** [comment]
> > Add a service field whenever emitting the account.signed event and only count towards DAU/MAU if service=sync.
>
> It's not obvious to me why this one is necessary, since DAU/MAU are aggregated by unique uid anyway I don't think it matters if we count both an account.signed and some oauth-token-related event for the same activity.

You are right, if we can merge the 3 distinct event types based on uid to count as a single DAU/MAU, then we won't need to.

**Comment by shane-tomlinson** [comment]
> Add a service field whenever emitting the account.signed event and only count towards DAU/MAU if service=sync.

`service` is already being emit with the `fxa_activity - cert_signed` event, see:

<img width="961" alt="Screenshot 2019-03-27 at 11 27 53" src="https://user-images.githubusercontent.com/848085/55072579-665fc100-5083-11e9-8469-99be4a5b8b7f.png">

**Comment by shane-tomlinson** [comment]
@irrationalagent - to log the new events to amplitude, I'll need to know the new event names and event properties. Do you want a [new group added to fxa-shared](https://github.com/mozilla/fxa-shared/blob/858be5242b86f3fcd15d6277f77f1019db5a0410/metrics/amplitude.js#L17), or like the last events, [use fxa_notify](https://github.com/mozilla/fxa-shared/blob/858be5242b86f3fcd15d6277f77f1019db5a0410/metrics/amplitude.js#L23), or maybe [fxa_activity](https://github.com/mozilla/fxa-shared/blob/master/metrics/amplitude.js#L18) makes more sense?

The two event names we'll need:

* When an access token is created
* When an access token's verification status is checked

@rfk, I only just saw this question here:

> Do we need to associate metrics context with these tokens on an ongoing basis, or is it a once-of thing for the purposes of tracking flow completion?

@philbooth and I were discussing this today in IRC, for the immediate needs here no, long term it would be desirable. The metricsContext is used by the auth-server to get deviceId, flowId, flowBeginTime, and the service name. If the goal is to count MAU, we need the uid and some other metadata which we can get using the access/refresh token and request headers. metricsContext isn't strictly necessary for MAU, so I'm going to punt on that and file a follow on so there is at least a chance of getting this into train-134.

**Comment by irrationalagent** [comment]
> @irrationalagent - to log the new events to amplitude, I'll need to know the new event names and event properties. Do you want a [new group added to fxa-shared](https://github.com/mozilla/fxa-shared/blob/858be5242b86f3fcd15d6277f77f1019db5a0410/metrics/amplitude.js#L17), or like the last events, [use fxa_notify](https://github.com/mozilla/fxa-shared/blob/858be5242b86f3fcd15d6277f77f1019db5a0410/metrics/amplitude.js#L23), or maybe [fxa_activity](https://github.com/mozilla/fxa-shared/blob/master/metrics/amplitude.js#L18) makes more sense?

I think that fxa_activity makes the most sense, given that they are going to be serving a similar purpose analysis-wise to cert_signed. 

> 
> The two event names we'll need:
> 
>     * When an access token is created
> 
>     * When an access token's verification status is checked

`fxa_activity - access_token_created`, `fxa_activity - access_token_checked` are fine by me.

**Comment by davismtl** [comment]
> Another thing to watch out for would be OAuth RPs that are using the tokens for their own purpoes (such as checking if their cache of the user's profile data is up-to-date) that does not actually correspond to activity from the user. Not entirely sure what to do about that though, apart from telling them not to do that :-)

Is there a message we should prepare for the RPs to communicate our recommendations? Is there a best practice we should be adding to our docs?

**Comment by shane-tomlinson** [comment]
> Another thing to watch out for would be OAuth RPs that are using the tokens for their own purpoes (such as checking if their cache of the user's profile data is up-to-date) that does not actually correspond to activity from the user.

I thought we actually *recommended* RPs do this on occasion (at least that's what the RP checklist suggests)? Is there another way for an RP that's not hooked up receive `profileDataChanged` notifications to ensure their data is up to date?
